File: hermes-runtime/hermes_runtime/datastore/handlers/dead_letter.py
"""Datastore handlers for ``toee_job_queue`` -- the dead-letter operator view and
governed Replay (0.0.4 S05, FR-13).

S01 gave the queue a terminal ``dead`` status, S02/S04 put every async trigger on
it, S03 made outbound sends idempotent. This is the human end of that: a
supervisor sees what is stuck and can put ONE job back on the queue, attributed
and audited.

**Two lists, because a dead job is not the only way work goes wrong.**

``jobs`` is FR-13's core -- ``status = 'dead'`` rows with type, payload summary,
attempts, ``last_error`` and timestamps, each joined to the ``outbound_send``
record it left behind (S03's ``idx_outbound_send_job``, which exists for exactly
this question: did this job already text the customer?).

``outbound`` is the states S03/S04 leave that NO dead-letter row captures, split
into buckets an operator can act on differently:

===================  ==========================================================
``send_failed``      The provider refused. The customer got nothing and nothing
                     will retry -- a reply, or (``slot = 'opt-out'``) the
                     ADR-0016 confirmation a customer who opted out never
                     received. A human decides whether to text them by hand.
``mirror_missing``   ``sent`` with a ``last_error``: the customer HAS the SMS
                     but the ``message_turn`` mirror never landed, so the
                     workbench thread is permanently incomplete. Do NOT re-send.
``stale_intent``     An ``intent`` row older than :data:`_STALE_INTENT_SECONDS`.
                     A process died between recording intent and the POST; the
                     re-run skipped quietly and **the job succeeded**, so this
                     is invisible everywhere else in the system. Nothing sweeps
                     these (deliberately -- S05 surfaces, it does not sweep).
===================  ==========================================================

Both actions are admin-only (``_AGENT_EXCLUDED_ACTIONS``) -- reached only from
the admin BFF's deterministic ``tools:dispatch`` call, never a live agent's tool
loop. The toolset is allowlisted on ``supervisor_admin`` (ADR-0038): this is an
operations surface, so it is supervisor+admin in the workbench (ADR-0093 gives
``/admin/*`` exactly that), deliberately unlike a credential surface.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from psycopg.rows import dict_row
from toee_hermes.errors import ToolDriverError

from ._common import insert_audit, read_string, serialize_row

if TYPE_CHECKING:  # pragma: no cover - typing only
    from toee_hermes.tool_gate import ToolExecutionContext

# ponytail: no pagination and no filters -- the whole point of a dead-letter view
# is that it is normally empty, and 100 stuck jobs is already an incident rather
# than a page-2 problem. Add a cursor the day an operator hits the cap.
_LIST_LIMIT = 100

# How old an `intent` row must be before it is a stuck send rather than one in
# flight. Sized at 3x DEFAULT_LEASE_SECONDS (300s): a turn that is still legally
# running holds a lease, and one lease plus its retry backoff is the longest a
# healthy `intent` row can honestly live.
# ponytail: a constant, not a per-channel policy. Make it configurable the day a
# channel's delivery is genuinely slower than an SMS POST.
_STALE_INTENT_SECONDS = 900

# A payload value's cap in the triage list. `l6_review` carries the entire review
# prompt (draft text included); the view is a list, not a transcript.
_SUMMARY_CHARS = 160

_JOB_COLUMNS = (
    "j.id, j.type, j.payload, j.attempts, j.max_attempts, j.last_error,"
    " j.run_at, j.created_at, j.updated_at"
)


def _summarize(payload: Any) -> dict[str, Any]:
    """Keys with short values -- long strings truncated, containers collapsed."""
    if not isinstance(payload, dict):
        return {}
    summary: dict[str, Any] = {}
    for key, value in payload.items():
        if isinstance(value, str) and len(value) > _SUMMARY_CHARS:
            summary[key] = value[:_SUMMARY_CHARS] + "…"
        elif isinstance(value, (dict, list)):
            summary[key] = f"<{type(value).__name__} of {len(value)}>"
        else:
            summary[key] = value
    return summary


def _list_dead_letters(
    conn, params: dict[str, Any], context: "ToolExecutionContext"
) -> Any:
    # A read -> no actor required, no audit (parity with get_retention_status).
    del params, context
    from hermes_runtime.job_queue import replay_blocked_reason

    with conn.cursor(row_factory=dict_row) as cur:
        cur.execute(
            f"""
            SELECT {_JOB_COLUMNS},
                   o.status AS outbound_status,
                   o.skip_count AS outbound_skip_count,
                   o.last_error AS outbound_last_error
            FROM job j
            LEFT JOIN outbound_send o ON o.job_id = j.id
            WHERE j.status = 'dead'
            ORDER BY j.updated_at DESC
            LIMIT {_LIST_LIMIT}
            """  # noqa: S608 - both interpolations are module constants
        )
        job_rows = [serialize_row(row) for row in cur.fetchall()]

        cur.execute(
            """
            SELECT idempotency_key, job_id, event_id, conversation_id, channel,
                   status, skip_count, last_error, created_at, updated_at,
                   regexp_replace(idempotency_key, '^.*:', '') AS slot
            FROM outbound_send
            WHERE status = 'failed'
               OR (status = 'sent' AND last_error IS NOT NULL)
               OR (status = 'intent'
                   AND updated_at < now() - make_interval(secs => %s))
            ORDER BY updated_at DESC
            LIMIT %s
            """,
            (_STALE_INTENT_SECONDS, _LIST_LIMIT),
        )
        outbound_rows = [serialize_row(row) for row in cur.fetchall()]

    jobs = []
    for row in job_rows:
        blocked = replay_blocked_reason(row["type"])
        outbound_status = row["outbound_status"]
        jobs.append(
            {
                "job_id": row["id"],
                "type": row["type"],
                "payload_summary": _summarize(row["payload"]),
                "attempts": row["attempts"],
                "max_attempts": row["max_attempts"],
                "last_error": row["last_error"],
                "run_at": row["run_at"],
                "created_at": row["created_at"],
                "updated_at": row["updated_at"],
                "replayable": blocked is None,
                "replay_blocked_reason": blocked,
                "outbound": (
                    None
                    if outbound_status is None
                    else {
                        "status": outbound_status,
                        "skip_count": row["outbound_skip_count"],
                        "last_error": row["outbound_last_error"],
                    }
                ),
            }
        )

    outbound = [{**row, "bucket": _bucket(row)} for row in outbound_rows]
    return {"jobs": jobs, "outbound": outbound}


def _bucket(row: dict[str, Any]) -> str:
    if row["status"] == "failed":
        return "send_failed"
    if row["status"] == "sent":
        return "mirror_missing"
    return "stale_intent"


def _require_actor(context: "ToolExecutionContext") -> str:
    """The acting supervisor for a governed replay, or a denial (ADR-0148).

    The actor rides ``ToolExecutionContext.user_id``, asserted by the BFF from
    the signed-in session under the shared bearer -- NEVER a request param.
    Fail closed: an unattributed replay would put work back on the queue and
    write a NULL-actor audit row while reporting success. (Mirrors
    ``knowledge._require_actor``.)
    """
    actor = context.user_id
    if not actor:
        raise ToolDriverError(
            "policy_blocked", "A governed job replay requires an attributed actor."
        )
    return actor


def _replay_job(conn, params: dict[str, Any], context: "ToolExecutionContext") -> Any:
    """Return one dead job to the queue, attributed and audited (FR-13).

    Order matters: actor first (fail closed), then the type check, then the
    reset -- so a blocked or unattributed replay leaves both the ``job`` row and
    the audit log untouched. The reset and the audit row share the handler's
    transaction (``PostgresDriver.execute`` commits around the whole call,
    ADR-0140), so a replay is never recorded without happening.

    No bulk replay in v1 (PRD default): one ``job_id``, one row. That is also
    what keeps ``ingest`` safe -- see ``REPLAY_BLOCKED_JOB_TYPES``.
    """
    from hermes_runtime.job_queue import replay_blocked_reason, replay_dead_job

    job_id = read_string(params, "job_id", "jobId")
    if not job_id:
        raise ToolDriverError("unexpected_error", "job_id is required.")
    actor = _require_actor(context)

    with conn.cursor() as cur:
        cur.execute("SELECT type, attempts FROM job WHERE id = %s AND status = 'dead'", (job_id,))
        row = cur.fetchone()
        if row is None:
            raise ToolDriverError("not_found", f"no dead job {job_id}.")
        job_type, attempts_before = row

        blocked = replay_blocked_reason(job_type)
        if blocked is not None:
            raise ToolDriverError("policy_blocked", blocked)

        replayed_type = replay_dead_job(cur, job_id)
        if replayed_type is None:  # raced another replay between the two statements
            raise ToolDriverError("not_found", f"no dead job {job_id}.")

    insert_audit(
        conn,
        profile=context.profile,
        account_id=actor,
        action="job_replayed",
        target_type="job",
        target_id=job_id,
        details={"type": job_type, "attempts_before": attempts_before},
    )
    return {"job_id": job_id, "type": job_type, "status": "queued"}


def dead_letter_handlers() -> dict[str, dict[str, Any]]:
    """Registry fragment for the dead-letter view + governed replay."""
    return {
        "toee_job_queue": {
            "list_dead_letters": _list_dead_letters,
            "replay_job": _replay_job,
        }
    }


__all__ = ["dead_letter_handlers"]

File: apps/workbench/lib/bff/admin/dead-letter.ts
// Dead-letter operator view + governed Replay (0.0.4 S05, FR-13) for the Admin
// BFF (ADR-0093 admin route group). Pure and dependency-injected like the sibling
// admin/*.ts modules; the thin app/api/admin/dead-letter* routes wrap this with
// withSession + createAdminApiClient.
//
// Dispatches over the SUPERVISOR ADMIN Profile API (deps.ts's
// createAdminApiClient), not the Internal Copilot one admin/retention.ts and
// admin/metrics.ts use: toee_job_queue is allowlisted for supervisor_admin
// (hermes/toee_hermes/plugin/profiles.py). Both actions are also admin-only on
// the Hermes side (_AGENT_EXCLUDED_ACTIONS) -- never reachable from a live
// agent's tool loop.
//
// ROLE GATING: /admin/* is supervisor+admin (lib/auth/access.ts). That is the
// right level for an OPERATIONS surface and is deliberately broader than a
// credential surface would be -- a supervisor triaging stuck work should not need
// an admin account, while credentials should.
//
// list is READ, fail-open (dispatch, not dispatchWrite) -- same convention as
// handleGetRetentionStatusViaApi. Replay is a governed WRITE (it puts work back
// on the queue): dispatchWrite, fail-closed on a missing actor. The acting
// account rides HermesApiClient.actorAccountId from the signed-in session
// (ADR-0148) -- it is never a request param, and the route never reads one.
import type { HermesApiClient } from "../../gateway/hermes-api-client";
import { HermesApiError } from "../../gateway/hermes-api-client";
import { hermesErrorToProblem } from "../../gateway/hermes-error";
import { json } from "../respond";

// What a dead job's own outbound attempt left behind (S03). `null` when the job
// never reached delivery -- which is what tells an operator a replay will
// genuinely send.
export interface DeadJobOutbound {
  status: string;
  skipCount: number;
  lastError: string | null;
}

export interface DeadJob {
  jobId: string;
  type: string;
  payloadSummary: Record<string, unknown>;
  attempts: number;
  maxAttempts: number;
  lastError: string | null;
  runAt: string | null;
  createdAt: string | null;
  updatedAt: string | null;
  // Per-job-type replay safety, decided server-side (job_queue.py's
  // REPLAY_BLOCKED_JOB_TYPES). The panel disables the button and shows the
  // reason; the handler denies the call regardless, so a stale list cannot
  // sneak an l6_review replay through.
  replayable: boolean;
  replayBlockedReason: string | null;
  outbound: DeadJobOutbound | null;
}

// An outbound_send row that needs a human although no dead-letter row exists.
// `bucket` is one of send_failed | mirror_missing | stale_intent -- see the
// handler docstring for what each means and what an operator should do.
export interface StuckOutbound {
  bucket: string;
  slot: string;
  idempotencyKey: string;
  jobId: string | null;
  eventId: string;
  conversationId: string;
  channel: string;
  status: string;
  skipCount: number;
  lastError: string | null;
  createdAt: string | null;
  updatedAt: string | null;
}

// One `job_replayed` audit row, named to an account. FR-13 asks for a VISIBLE
// audit row; the handler writes it with target_type='job' and every other
// workbench audit view is case- or record-scoped, so this short tail is where a
// replay's provenance can actually be seen (and the only place it survives a
// replay that SUCCEEDS and takes its job off the dead list).
export interface RecentReplay {
  jobId: string;
  type: string | null;
  accountId: string | null;
  actorUsername: string | null;
  createdAt: string | null;
}

export interface DeadLetterView {
  jobs: DeadJob[];
  outbound: StuckOutbound[];
  recentReplays: RecentReplay[];
}

export interface ReplayReceipt {
  jobId: string;
  type: string | null;
  status: string;
}

function malformed(detail: string): never {
  throw new HermesApiError(
    "unexpected_error",
    `malformed dead-letter payload: ${detail}`,
  );
}

function requireString(value: unknown, field: string): string {
  if (typeof value !== "string") malformed(field);
  return value as string;
}

function requireNullableString(value: unknown, field: string): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value !== "string") malformed(field);
  return value as string;
}

function requireNumber(value: unknown, field: string): number {
  if (typeof value !== "number" || Number.isNaN(value)) malformed(field);
  return value as number;
}

function requireObject(value: unknown, field: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    malformed(field);
  }
  return value as Record<string, unknown>;
}

function requireArray(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) malformed(field);
  return value as unknown[];
}

function mapOutboundOnJob(value: unknown): DeadJobOutbound | null {
  if (value === null || value === undefined) return null;
  const r = requireObject(value, "outbound");
  return {
    status: requireString(r.status, "outbound.status"),
    skipCount: requireNumber(r.skip_count, "outbound.skip_count"),
    lastError: requireNullableString(r.last_error, "outbound.last_error"),
  };
}

function mapDeadJob(raw: unknown): DeadJob {
  const r = requireObject(raw, "jobs[]");
  return {
    jobId: requireString(r.job_id, "job_id"),
    type: requireString(r.type, "type"),
    payloadSummary: requireObject(r.payload_summary ?? {}, "payload_summary"),
    attempts: requireNumber(r.attempts, "attempts"),
    maxAttempts: requireNumber(r.max_attempts, "max_attempts"),
    lastError: requireNullableString(r.last_error, "last_error"),
    runAt: requireNullableString(r.run_at, "run_at"),
    createdAt: requireNullableString(r.created_at, "created_at"),
    updatedAt: requireNullableString(r.updated_at, "updated_at"),
    // Strict on purpose: a missing `replayable` must NOT default to true, or a
    // backend shape change would silently re-open replay for l6_review.
    replayable: typeof r.replayable === "boolean" ? r.replayable : malformed("replayable"),
    replayBlockedReason: requireNullableString(
      r.replay_blocked_reason,
      "replay_blocked_reason",
    ),
    outbound: mapOutboundOnJob(r.outbound),
  };
}

function mapStuckOutbound(raw: unknown): StuckOutbound {
  const r = requireObject(raw, "outbound[]");
  return {
    bucket: requireString(r.bucket, "bucket"),
    slot: requireString(r.slot, "slot"),
    idempotencyKey: requireString(r.idempotency_key, "idempotency_key"),
    jobId: requireNullableString(r.job_id, "job_id"),
    eventId: requireString(r.event_id, "event_id"),
    conversationId: requireString(r.conversation_id, "conversation_id"),
    channel: requireString(r.channel, "channel"),
    status: requireString(r.status, "status"),
    skipCount: requireNumber(r.skip_count, "skip_count"),
    lastError: requireNullableString(r.last_error, "last_error"),
    createdAt: requireNullableString(r.created_at, "created_at"),
    updatedAt: requireNullableString(r.updated_at, "updated_at"),
  };
}

function mapRecentReplay(raw: unknown): RecentReplay {
  const r = requireObject(raw, "recent_replays[]");
  return {
    jobId: requireString(r.job_id, "recent_replays[].job_id"),
    type: requireNullableString(r.type, "recent_replays[].type"),
    accountId: requireNullableString(r.account_id, "recent_replays[].account_id"),
    actorUsername: requireNullableString(
      r.actor_username,
      "recent_replays[].actor_username",
    ),
    createdAt: requireNullableString(r.created_at, "recent_replays[].created_at"),
  };
}

export function mapDeadLetterView(raw: unknown): DeadLetterView {
  const r = requireObject(raw, "root");
  return {
    jobs: requireArray(r.jobs, "jobs").map(mapDeadJob),
    outbound: requireArray(r.outbound, "outbound").map(mapStuckOutbound),
    // Deliberately NOT strict like `replayable`: this list is provenance
    // display, and a backend that omits it cannot re-open a blocked action --
    // whereas a defaulted `replayable` would. A backend without the field (the
    // mock driver) shows an empty log rather than 502-ing the whole panel.
    recentReplays: requireArray(
      r.recent_replays ?? [],
      "recent_replays",
    ).map(mapRecentReplay),
  };
}

export function mapReplayReceipt(raw: unknown): ReplayReceipt {
  const r = requireObject(raw, "root");
  return {
    jobId: requireString(r.job_id, "job_id"),
    type: requireNullableString(r.type, "type"),
    status: requireString(r.status, "status"),
  };
}

export async function handleListDeadLettersViaApi(
  client: HermesApiClient,
): Promise<Response> {
  try {
    const data = await client.dispatch("toee_job_queue", "list_dead_letters", {});
    return json(mapDeadLetterView(data));
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export async function handleReplayJobViaApi(
  client: HermesApiClient,
  jobId: unknown,
): Promise<Response> {
  if (typeof jobId !== "string" || jobId === "") {
    return hermesErrorToProblem(
      new HermesApiError("not_found", "jobId is required"),
    );
  }
  try {
    // No bulk replay in v1 (PRD default): one job id, one call. The actor is NOT
    // in these params -- it rides the client's actorAccountId (ADR-0148).
    const data = await client.dispatchWrite("toee_job_queue", "replay_job", {
      job_id: jobId,
    });
    return json(mapReplayReceipt(data));
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}
